RFC-002-mapeo-protocol.md

```diff
@@ -0,0 +1,274 @@
+change over time, since if the entire protocol were encrypted, a breaking
+change would also break any ability for communication between peers across
+versions, rendering the peer-to-peer update feature useless.
+
+## 3.2 Secure Channel
+
+This is the channel that is secured using a cryptographic handshake to derive a
+shared ephemeral key (via [NOISE][]), and a bulk encryption algorithm to
+encrypt all of the outgoing data and decrypt incoming data (via [XSalsa20][]).
+
+[NOISE][] was chosen for its properties of providing end-to-end encryption, as
+well as forward secrecy. [Secret Handshake][] was also reviewed, but the
+requirement of knowing each peer's public key in advance did not fit with
+Mapeo's current discovery model, where peers do not need to be known in advance
+to be found on the local network.
+
+[XSalsa20][] was chosen because of its ongoing use in the [hypercore][]
+project, which Mapeo currently depends on for encryption of the hypercore
+protocol.
+
+An encrypted RPC protocol will run over this secure channel, which can support
+subprotocols, such as Hypercore sync, or media blob exchange. All information
+between peers that requires privacy will be passed through this channel.
+
+Finally, if the secure channel cannot be established due to incompatibility
+reasons (such as an update to another encryption algorithm), it should still
+permit the plaintext channel to be established and used. However, if the secure
+channel fails to establish because of security reasons (eg. detecting a
+man-in-the-middle attack), it should abort the connection entirely.
+
+## 3.3 Remote Procedure Call (RPC) System
+
+The [muxrpc][] protocol wire format is documented on the
+[Secure Scuttlebutt Protocol Guide][ssb-proto-rpc].
+
+# 4. Plaintext Channel RPCs
+
+A peer can use this channel to request a listing of app upgrades (not limited
+to its own platform/architecture), download app binaries over it, get peer
+info, and open a secure channel.
+
+This happens on this unencrypted channel so that if/when we need to change the
+encryption protocol or how upgrades are distributed, the upgrade protocol will
+remain unaffected. The goal is to ensure that a very old version of Mapeo can
+still interoperate at the App Upgrade Protocol layer, even if the secure
+protocol's shape has diverged significantly.
+
+Further, plaintext is sufficiently secure for our needs, since application
+binaries are already signed by default for MacOS, Windows, and Android. For
+Linux, an ad-hoc signing mechanism will need to be deployed. The digital
+signature mechanism prevents a man-in-the-middle from modifying the payload
+undetected. The listing of available upgrades will also include a hash, for
+detecting transmission errors.
+
+The upgrade protocol is expressed as an RPC protocol for requests and responses
+between peers. There are three RPCs exposed by this channel:
+
+## 4.1 `GetAvailableBinaries()`
```

Comment:
There is also the possibility that we could distribute OTA JS bundles for updates. On mobile this is possible to implement using [expo-updates](https://docs.expo.io/versions/latest/sdk/updates/). Any release that only included changes to React Native JS code could be updated via a JS bundle (which would be about 1-2Mb vs. 50Mb for the whole APK). It is theoretically possible to also update the backend Node code on mobile this way too, but there is no existing library to manage this. It is also theoretically possible to update electron this way too (distribute only an updated JS bundle rather than the whole electron app) but I have not seen any attempts to write this. In general the electron community seems to be resistant even to attempts to reduce the download size since the core developers do not see ~50-100Mb as significant for downloads.

How do you think we might expose these updates in the future in a backwards compatible way?

On Mobile, if we start using `expo-updates` to distribute updates to phones with internet access, then the APK on the phone (that can be shared and installed) will not include the updated JS bundle (it is downloaded to disk, and loaded at runtime). This would mean that an update would require both the APK + the JS bundle.

We will not be implementing this feature (OTA JS bundle updates) in the short term, but it would be good to think about how we might implement it in a non-breaking way in the future.

Reply:
With electron updater you can update using 'blockmap' which is not the full app, only the pieces that have changed.

This seems backwards compatible as-is, since we can always add a new RPC call for a particular type of update

Reply:
Do the JS bundles need to assume the recipient already has a certain version range already? My understanding is that you can't do a fresh install from just a JS bundle, and that sometimes we'll release versions with changes to non-JS bits. If that's true, we could capture these types of updates with a `requiredMinVersion` property, where the recipient would not download an update with `{ version: "5.0.0", requiredMinVersion: "4.1.3" }` if they had, e.g. `"4.1.2"`. Alternatively, we could have a version scheme that incorporates JS bundles, e.g. `"4.1.5-js6"` to indicate that this is the 6th JS bundle patch for 4.1.5. What do y'all think?

Also, if we need special handling for these kinds of updates, we could add a field to the metadata that says whether it's a full drop-in update, or a special type with restrictions (OTA JS, expo, electron blockmap) so clients can implement whatever handling they need. This will also let us figure out the necessary fields at a later time instead of right now.

Reply:
Yeah, I agree we can figure out the necessary fields at a later time. The OTA updates would require a version range. We would also likely be publishing new APKs with the updated JS bundle included in the APK.

Reply:
Ok! I added a note about this in the RFC so we don't forget.